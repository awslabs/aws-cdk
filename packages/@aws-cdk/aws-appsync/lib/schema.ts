import { IBucket } from '@aws-cdk/aws-s3';
import { CfnGraphQLSchema } from './appsync.generated';
import { GraphqlApi } from './graphqlapi';
import { AssetSchema, CodeSchema, S3Schema } from './private/schema-implementation';
import { IIntermediateType } from './schema-base';
import { ResolvableField } from './schema-field';
import { ObjectType } from './schema-intermediate';

/**
 * The Schema for a GraphQL Api
 *
 * If no options are configured, schema will be generated
 * code-first.
 */
export abstract class Schema {
  /**
   * Generate a Schema from file
   *
   * @returns `Schema` with immutable schema defintion
   * @param filePath the file path of the schema file
   */
  public static fromAsset(filePath: string): Schema {
    return new AssetSchema(filePath);
  }

  /**
   * Point GraphQL Schema to a S3 Location
   *
   * @returns `Schema` with immutable schema defintion
   * @param bucket the s3 bucket that holds the GraphQL Schema
   * @param key the object key for the schema
   */
  public static fromBucket(bucket: IBucket, key: string): Schema {
    return new S3Schema(bucket, key);
  }

  /**
   * Generate a Schema that can be edited through a code-first
   * approach
   *
   * @returns `Schema` with mutable schema defintion
   */
  public static fromCode(): Schema {
    return new CodeSchema();
  }

  /**
   * The definition for this schema
   */
  public readonly abstract definition: string;

  /**
   * Called when the GraphQL Api is initialized to allow this object to bind
   * to the stack.
   */
  public abstract bind(api: GraphqlApi): CfnGraphQLSchema;

  /**
   * Escape hatch to add to Schema as desired. Will always result
   * in a newline.
   */
  public abstract addToSchema(addition: string, delimiter?: string): void;
  /**
   * Add a query field to the schema's Query. CDK will create an
   * Object Type called 'Query'. For example,
   *
   * type Query {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addQuery(fieldName: string, field: ResolvableField): ObjectType;

  /**
   * Add a mutation field to the schema's Mutation. CDK will create an
   * Object Type called 'Mutation'. For example,
   *
   * type Mutation {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addMutation(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add a subscription field to the schema's Subscription. CDK will create an
   * Object Type called 'Subscription'. For example,
   *
   * type Subscription {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addSubscription(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add type to the schema
   */
  public abstract addType(_type: IIntermediateType): IIntermediateType;
}