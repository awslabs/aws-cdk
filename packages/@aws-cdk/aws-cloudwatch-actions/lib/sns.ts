import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import * as sns from '@aws-cdk/aws-sns';
import { Construct } from '@aws-cdk/core';

/**
 * Use an SNS topic as an alarm action
 */
export class SnsAction implements cloudwatch.IAlarmAction {
  constructor(private readonly topic: sns.ITopic) {
  }

  /**
   * Returns an alarm action configuration to use an SNS topic as an alarm action
   */
  public bind(_scope: Construct, _alarm: cloudwatch.IAlarm): cloudwatch.AlarmActionConfig {
    if (this.topic.masterKey) {
      this.topic.masterKey.grantEncryptDecrypt(new ServicePrincipal('cloudwatch.amazonaws.com'));
    }
    return { alarmActionArn: this.topic.topicArn };
  }
}
