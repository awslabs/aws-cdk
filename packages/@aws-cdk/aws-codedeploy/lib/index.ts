export * from './rollback-config';
export * from './ecs';
export * from './lambda';
export * from './server';
export * from './trigger-configuration';

// AWS::CodeDeploy CloudFormation Resources:
export * from './codedeploy.generated';
