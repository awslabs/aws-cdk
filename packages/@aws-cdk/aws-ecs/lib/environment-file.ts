import { IBucket, Location } from '@aws-cdk/aws-s3';
import * as s3_assets from '@aws-cdk/aws-s3-assets';
import { FileAsset, FileAssetOptions } from '@aws-cdk/core';
import { toSymlinkFollow } from './private/follow';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from '@aws-cdk/core';

/**
 * Constructs for types of environment files
 */
export abstract class EnvironmentFile {
  /**
   * Loads the environment file from a local disk path.
   *
   * @param path Local disk path
   * @param options
   *
   * @deprecated use fromFileAsset instead
   */
  public static fromAsset(path: string, options?: s3_assets.AssetOptions): AssetEnvironmentFile {
    return new AssetEnvironmentFile(path, options);
  }

  /**
   * Loads the environment file from a local disk path.
   *
   * @param path Local disk path
   * @param options
   */
  public static fromFileAsset(path: string, options?: FileAssetOptions): EnvironmentFile {
    return new FileAssetEnvironmentFile(path, options);
  }

  /**
   * Loads the environment file from an S3 bucket.
   *
   * @returns `S3EnvironmentFile` associated with the specified S3 object.
   * @param bucket The S3 bucket
   * @param key The object key
   * @param objectVersion Optional S3 object version
   */
  public static fromBucket(bucket: IBucket, key: string, objectVersion?: string): S3EnvironmentFile {
    return new S3EnvironmentFile(bucket, key, objectVersion);
  }

  /**
   * Called when the container is initialized to allow this object to bind
   * to the stack.
   *
   * @param scope The binding scope
   */
  public abstract bind(scope: Construct): EnvironmentFileConfig;
}

/**
 * Environment file from a local directory.
 *
 * @deprecated use EnvironmentFile.fromFileAsset instead
 */
export class AssetEnvironmentFile extends EnvironmentFile {
  private readonly fileAssetEnvrionmentFile: FileAssetEnvironmentFile;

  /**
   * @param path The path to the asset file or directory.
   * @param options
   */
  constructor(path: string, options: s3_assets.AssetOptions = { }) {
    super();

    this.fileAssetEnvrionmentFile = new FileAssetEnvironmentFile(path, {
      ...options,
      followSymlinks: options.followSymlinks ?? toSymlinkFollow(options.follow),
    });
  }

  public bind(scope: Construct): EnvironmentFileConfig {
    return this.fileAssetEnvrionmentFile.bind(scope);
  }
}

class FileAssetEnvironmentFile extends EnvironmentFile {
  private asset?: FileAsset;

  /**
   * @param path The path to the asset file or directory.
   * @param options
   */
  constructor(public readonly path: string, private readonly options: FileAssetOptions = { }) {
    super();
  }

  public bind(scope: Construct): EnvironmentFileConfig {
    // If the same AssetCode is used multiple times, retain only the first instantiation.
    if (!this.asset) {
      this.asset = new FileAsset(scope, 'EnvironmentFile', {
        path: this.path,
        ...this.options,
      });
    }

    if (!this.asset.isFile) {
      throw new Error(`Asset must be a single file (${this.path})`);
    }

    return {
      fileType: EnvironmentFileType.S3,
      s3Location: {
        bucketName: this.asset.s3BucketName,
        objectKey: this.asset.s3ObjectKey,
      },
    };
  }
}

/**
 * Environment file from S3.
 */
export class S3EnvironmentFile extends EnvironmentFile {
  private readonly bucketName: string;

  constructor(bucket: IBucket, private key: string, private objectVersion?: string) {
    super();

    if (!bucket.bucketName) {
      throw new Error('bucketName is undefined for the provided bucket');
    }

    this.bucketName = bucket.bucketName;
  }

  public bind(_scope: Construct): EnvironmentFileConfig {
    return {
      fileType: EnvironmentFileType.S3,
      s3Location: {
        bucketName: this.bucketName,
        objectKey: this.key,
        objectVersion: this.objectVersion,
      },
    };
  }
}

/**
 * Configuration for the environment file
 */
export interface EnvironmentFileConfig {
  /**
   * The type of environment file
   */
  readonly fileType: EnvironmentFileType;

  /**
   * The location of the environment file in S3
   */
  readonly s3Location: Location;
}

/**
 * Type of environment file to be included in the container definition
 */
export enum EnvironmentFileType {
  /**
   * Environment file hosted on S3, referenced by object ARN
   */
  S3 = 's3',
}
