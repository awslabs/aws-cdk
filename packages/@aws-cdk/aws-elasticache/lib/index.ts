// AWS::ElastiCache CloudFormation Resources:
export * from './elasticache.generated';
export * from './parameter-group';
