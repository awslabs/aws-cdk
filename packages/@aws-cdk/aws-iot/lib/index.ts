export * from './topic-rule';
export * from './topic-rule-action';
// AWS::IoT CloudFormation Resources:
export * from './iot.generated';
