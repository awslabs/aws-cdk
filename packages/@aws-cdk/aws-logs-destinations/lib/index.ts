export * from './lambda';
export * from './kinesis';
export * from './kinesisfirehose';
