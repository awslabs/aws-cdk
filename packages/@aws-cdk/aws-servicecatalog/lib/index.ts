export * from './common';
export * from './constraints';
export * from './cloudformation-template';
export * from './portfolio';
export * from './product';
export * from './tagoptions';

// AWS::ServiceCatalog CloudFormation Resources:
export * from './servicecatalog.generated';
