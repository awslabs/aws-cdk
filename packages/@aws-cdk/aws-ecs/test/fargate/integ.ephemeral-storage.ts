import * as cdk from '@aws-cdk/core';
import * as ecs from '../../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-ecs-ephemeral-storage');

const taskDefinition = new ecs.FargateTaskDefinition(stack, 'TaskDef', {
  ephemeralStorageGiB: 100,
});

taskDefinition.addContainer('web', {
  image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
});

app.synth();
