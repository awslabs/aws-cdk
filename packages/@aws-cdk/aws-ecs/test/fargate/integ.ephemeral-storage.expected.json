{
  "Resources": {
    "TaskDefTaskRole1EDB4A67": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "TaskDef54694570": {
      "Type": "AWS::ECS::TaskDefinition",
      "Properties": {
        "ContainerDefinitions": [
          {
            "Essential": true,
            "Image": "amazon/amazon-ecs-sample",
            "Name": "web"
          }
        ],
        "Cpu": "256",
        "EphemeralStorage": {
          "SizeInGiB": 100
        },
        "Family": "awsecsephemeralstorageTaskDef7E6908C5",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE"
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "TaskDefTaskRole1EDB4A67",
            "Arn"
          ]
        }
      }
    }
  }
}